### ISSUE TITLE
About measuring personalization 

### ISSUE THREAD

**Issue by zhunipingan** [majumderb/recipe-personalization#6]
Hi,
   Thanks for your sharing. I have questions to ask you. How is the personalization evaluation calculated. Is it to pre train the evaluation model according to the training set, and then use the pre-trained model to evaluate the personalized performance in the evaluation part. And I can't find the code about this part. Can you tell me which file is related to the evaluation.
    Thanks!

**Comment by majumderb** [comment]
It is a ranking based metric. As we have described in the paper, you need to create 10 pairs for each input such as: one pair is (input, correct output), and the rest are (input, incorrect output). In our case, the correct output is the original recipe and the incorrect output is any random recipe. 

Then you need to use the recipe generation model you have trained to calculate the conditional likelihood of each pair and rank them accordingly. After ranking, look for the rank of the correct pair in the list of 10 pairs and calculate accuracy/MRR as usual.

**Comment by zhunipingan** [comment]
Thanks. But how the likelihood is calculated. In my view, the recipe genration model only can output one long sentence(recipe), but no likelihood is outputed. 
The one of 10 pairs is as ((recipe name, ingredients, **correct user profile**), correct recipe), and the rest 9 is ((recipe name, ingredients, **incorrect user profile**), correct recipe), but how to calculate how closely correct recipe corresponds to a particular user profile. I am still don't understand.

**Comment by shuyangli94** [comment]
In order to create recipe words, we sample from the logits/softmax probabilities at each time step, which are returned [here](https://github.com/majumderb/recipe-personalization/blob/master/recipe_gen/models/user_item_pref/__init__.py#L308).

You can measure likelihood via measures like the mean log likelihood of the chosen token at each time step. To compare recipes, we rank the 10 generated recipes by likelihood and then find the rank of the recipe generated by the correct user profile. You can then measure HR@1, MRR, etc. using that rank value. The intuition is that the generated recipe should be more likely if conditioned on the correct user.
